### PR TITLE
Feat/icon button as column type

### LINKED ISSUE #5635: Include icon button as column type
# Objective
We will introduce the icon button in the table column, Follow all properties of the button column type and the icon button widget.

# Requirement
| Title | Desc | Comment | 
| ------ | ------- | ------ |
| Select icon | The default icon control to search and select | |
| icon styles | As described in the icon button | |
| Radius | " | | 
| Shadow | " | |
| others | All other properties from the button column | |

### PR DESCRIPTION
**Pull request by Tooluloope**
## Description

> Add IconButton as column type

Fixes #5635 

## Type of change

- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Test coverage results :test_tube:
<details><summary>:green_circle: Total coverage has increased</summary>

    // Code coverage diff between base branch:release and head branch: feat/icon-button-as-column-type 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :green_circle: | total | 54.92 **(0)** | 36.8 **(-0.02)** | 33.69 **(0.05)** | 55.46 **(0)**
 :red_circle: | app/client/src/components/designSystems/appsmith/IconButtonComponent.tsx | 50 **(0)** | 8 **(-0.33)** | 22.22 **(0)** | 60.61 **(0)**
 :red_circle: | app/client/src/components/designSystems/appsmith/TableComponent/TableUtilities.tsx | 39.16 **(-3.93)** | 18.45 **(-1.14)** | 15.79 **(-2.96)** | 37.69 **(-3.9)**
 :red_circle: | app/client/src/components/propertyControls/ColumnActionSelectorControl.tsx | 40.43 **(0)** | 28.57 **(-3.01)** | 16.67 **(0)** | 42.86 **(0)**
 :green_circle: | app/client/src/utils/autocomplete/TernServer.ts | 50.71 **(0.24)** | 38.6 **(0.88)** | 36.21 **(0)** | 55.35 **(0.27)**
 :green_circle: | app/client/src/widgets/TableWidget/TablePropertyPaneConfig.ts | 38.73 **(10.04)** | 10.2 **(2.83)** | 65.63 **(18.01)** | 40.15 **(10.24)**
 :red_circle: | app/client/src/widgets/TableWidget/index.tsx | 27.14 **(-0.27)** | 5.92 **(-0.16)** | 25 **(-0.49)** | 27.27 **(-0.3)**</details>

### CODE REVIEW

File: app/client/src/components/designSystems/appsmith/TableComponent/TableUtilities.tsx [hunk 1]
```diff
@@ -132,6 +139,86 @@ export const renderCell = (
...
+  boxShadow: ButtonBoxShadow;
+  boxShadowColor: string;
+}): JSX.Element {
+  const [loading, setLoading] = useState(false);
+  const onComplete = () => {
+    setLoading(false);
+  };
+  return (
+    <div
+      onClick={(e) => {
```
Comment (marks0351) [inline app/client/src/components/designSystems/appsmith/TableComponent/TableUtilities.tsx:199]: pls declare the inline function.

File: app/client/src/components/propertyControls/ColumnActionSelectorControl.tsx [hunk 2]
```diff
@@ -52,13 +52,13 @@ class ColumnActionSelectorControl extends BaseControl<
                 <InputTextWrapper>
                   <InputText
                     evaluatedValue={columnAction.label}
-                    label={columnAction.label}
+                    label={columnAction.label as string}
```
Comment (marks0351) [inline app/client/src/components/propertyControls/ColumnActionSelectorControl.tsx:55]:
```suggestion
                    label={columnAction.label || ''}
```

File: app/client/src/widgets/TableWidget/TablePropertyPaneConfig.ts [hunk 3]
```diff
@@ -795,19 +891,159 @@ export default [
...
+                },
+                {
+                  propertyName: "boxShadow",
+                  label: "Box Shadow",
+                  helpText:
+                    "Enables you to cast a drop shadow from the frame of the widget",
+                  controlType: "BOX_SHADOW_OPTIONS",
+                  customJSControl: "COMPUTE_VALUE",
+                  isJSConvertible: true,
+                  hidden: (props: TableWidgetProps, propertyPath: string) => {
```
Comment (marks0351) [inline app/client/src/widgets/TableWidget/TablePropertyPaneConfig.ts:971]: is there a way to declare and reuse this function it seems to have been used in multiple places

File: app/client/src/widgets/TableWidget/TablePropertyPaneConfig.ts [hunk 4]
```diff
@@ -588,7 +592,8 @@ export default [
                 return (
```
Comment (marks0351) [inline app/client/src/widgets/TableWidget/TablePropertyPaneConfig.ts:592]: u cud create an array constant of all these types with a meaningful name which wud give more readability.
use the Array.includes function to check instead.

**Review by marks0351 — CHANGES_REQUESTED** [review]
Pls add jest or cypress tests as well.
Also do we need a docs PR update for this feature?

**Comment by Tooluloope** [comment]
> Pls add jest or cypress tests as well.
> Also do we need a docs PR update for this feature?

Screenshot 2021-08-22 at 09 36 38

@somangshu @marks0351 this is the only reference to columnType i saw on the docs.
Not sure  there's anywhere to update this on the docs

File: app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/DisplayWidgets/Table_Widget_Add_button_spec.js [hunk 5]
```diff
@@ -10,39 +10,56 @@ describe("Table Widget property pane feature validation", function() {
...
-    // Selecting button action to show message
-    cy.get(widgetsPage.actionSelect).click();
-    cy.get(commonlocators.chooseAction)
-      .children()
-      .contains("Show Message")
-      .click();
-    cy.addSuccessMessage("Successful ".concat(testdata.currentRowEmail));
-    // Close Property pane
-    cy.get(commonlocators.editPropCrossButton).click({ force: true });
+  // it("Table widget with Add button test and validation", function() {
```
Comment (marks0351) [inline app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/DisplayWidgets/Table_Widget_Add_button_spec.js:13]: why have we commented these tests?

**Review by marks0351 — CHANGES_REQUESTED** [review]

File: app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/DisplayWidgets/Table_Widget_Add_button_spec.js (hunk 5 above)
Comment (Tooluloope) [reply app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/DisplayWidgets/Table_Widget_Add_button_spec.js:13]: Fixed, I guess I skipped it during my review

**Review by marks0351 — APPROVED** [review]

**Comment by YogeshJayaseelan** [comment]
@marks0351 Tested this PR and please find the status below,

1.Click on any o the icon in the table throws the following 'Cannot read property 'replace' of undefined' 

https://www.loom.com/share/b2ca756a61f64451a813cd806707691b

**Comment by somangshu** [comment]
@YogeshJayaseelan please leave all feedback together if possible. This will help us minimize the to and fro

**Comment by Tooluloope** [comment]
@YogeshJayaseelan, Fixed the first issue, the same thing happens when you change to a Button type, You can open an issue for that and assign it to me

2. @somangshu, I felt it should have a JS option, since its on a Table, and Each row could have different icons and styles based on some other logic.

**Review by somangshu — DISMISSED** [review]
Approving to run tests

@marks0351 can you check the code and take a call once

**Review by marks0351 — DISMISSED** [review]

**Comment by YogeshJayaseelan** [comment]
@Tooluloope Verified the fix for 6th point and it is working as expected.

Suggestion:
7. Can we remove the icon columns in filter pane options as we ignored buttons in the filter pane option - clicking on icons in the filter pane throws - oops try again something went wrong error

https://www.loom.com/share/aa71b1d39a484bfcb97f7c16159eff81

**Comment by YogeshJayaseelan** [comment]
@Tooluloope Verified this PR and it is working as expected

**Review by somangshu — APPROVED** [review]
LGTM

**Review by marks0351 — APPROVED** [review]

**Pull request merged**